### ISSUE TITLE
Duplex streaming problems like: The client reset the request stream.

### ISSUE THREAD

**Issue by istvanek** [grpc/grpc-dotnet#1394]
Hello,
in our production application we have gRPC

- server using ASP.NET Core 5 implementation
- .NET Framework 4.8 clients using C# Grpc.Core implementation

Duplex streaming calls are used in various cases like
- notifications about data changes
- image streaming

Clients call the server, stream requests, read streamed responses and close request stream.

Sometimes we see a call ends with errors on server in requestStream.MoveNext, but has no problem on clients.
This is not problematic when we do all processing during streaming and error on server means actually a client has ended the call.
But this is problematic in cases when we do some processing on server after a client call end, but due to the error the processing is skipped and client just thinks everything is ok, which is not obviously.

To reproduce the problem, I have prepared testing service and client [attached to this post](https://github.com/grpc/grpc-dotnet/files/7076675/GrpcDuplexStreaming.zip):
- service uses ASP.NET Core 5 implementation
- client may choose .NET or Grpc.Core implementation, but both show the problem

Steps to reproduce the problem:
- start service
- start client and choose .NET or Grpc.Core implementation
- client calls the service duplex streaming call with short delay (100 ms which can be changed with +- keys) between responses
- service processes the call with short delay (100 ms which can be changed with +- keys) between responses
- to end the call on client and start another one, press any key (eg. space)
- depending on the call duration, the call ends well or with errors in the service - just press a key on client few times after various times
- to continue the service after any error, press Enter

The errors I have seen are:
- info: Grpc.AspNetCore.Server.ServerCallHandler[14]
      Error reading message.
      System.IO.IOException: The client reset the request stream.
         at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpRequestPipeReader.AdvanceTo(SequencePosition consumed, SequencePosition examined)
         at Grpc.AspNetCore.Server.Internal.PipeExtensions.ReadStreamMessageAsync[T](PipeReader input, HttpContextServerCallContext serverCallContext, Func`2 deserializer, CancellationToken cancellationToken)
System.IO.IOException: The client reset the request stream.
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpRequestPipeReader.AdvanceTo(SequencePosition consumed, SequencePosition examined)
   at Grpc.AspNetCore.Server.Internal.PipeExtensions.ReadStreamMessageAsync[T](PipeReader input, HttpContextServerCallContext serverCallContext, Func`2 deserializer, CancellationToken cancellationToken)
   at Grpc.AspNetCore.Server.Internal.HttpContextStreamReader`1.<MoveNext>g__MoveNextAsync|11_0(ValueTask`1 readStreamTask)
   at Service.GreeterService.SayHello(IAsyncStreamReader`1 requestStream, IServerStreamWriter`1 responseStream, ServerCallContext context) in GrpcDuplexStreaming\Service\Services\GreeterService.cs:line 21
- System.InvalidOperationException: Can't read messages after the request is complete.
   at Service.GreeterService.SayHello(IAsyncStreamReader`1 requestStream, IServerStreamWriter`1 responseStream, ServerCallContext context) in GrpcDuplexStreaming\Service\Services\GreeterService.cs:line 21

I consider this as severe problem and would like to see any fix or recommendation how to deal with it.

Thank You

Marek Ištvánek

**Comment by istvanek** [comment]
Can anybody have a look at it please?

**Comment by captainsafia** [comment]
Thanks for opening this issue, @istvanek! Have you tried reading messages to completion on the server side? You'll need to check for the value return from the `MoveNext` call per the directions [here](https://docs.microsoft.com/en-us/aspnet/core/grpc/client?view=aspnetcore-5.0#server-streaming-call).

**Comment by istvanek** [comment]
@captainsafia as You can see in the attached reproduction sample, I use it in the way described.
The problem lies in MoveNext on the server which just throws that exception instead of returning false.

**Comment by istvanek** [comment]
Any news?

**Comment by JamesNK** [comment]
Hi

I investigated your repo and I believe I found the problem.

Your code (I modified it slightly while debugging, but it is basically the same):
```cs
private static async Task Call(Greeter.GreeterClient client)
{
    callNumber++;
    Console.WriteLine($"{callNumber} Call");
    Console.WriteLine("Press any key for next call ...");
    uint number = 0;
    using var call = client.SayHello();
    while (true)
    {
        number++;
        HelloRequest request = new() { Message = number.ToString() };
        Console.WriteLine($"{callNumber} Request: {request.Message}");
        await call.RequestStream.WriteAsync(request);
        await call.ResponseStream.MoveNext();
        var response = call.ResponseStream.Current;
        Console.WriteLine($"{callNumber} Response: {response.Message}");
        if (!await ConsoleHelper.ApplyAndTryUpdateCallDelay())
        {
            Console.WriteLine($"{callNumber} Completing request");
            await call.RequestStream.CompleteAsync();
            break;
        }
    }
}
```

In the client you are calling `RequestStream.CompleteAsync()` and then immediately exiting the `Call` method. That causes the call to be disposed because of the `using`: `using var call = client.SayHello();`.

There appears to be a race in `Grpc.Net.Client` between the request stream gracefully completing from `CompleteAsync`, i.e. telling the server that the stream has ended, and it being terminated from the dispose. Because you are doing those two things immediately together then you see this issue. Adding a small `Task.Delay(10)` to the end of the `Call` method fixes it.

**Comment by JamesNK** [comment]
I think I found another issue on the server. It will require a fix in .NET which will take a while.

What I recommend you do is modify the client to gracefully complete both sides of the stream before disposing. That is best practice when doing streams.

To do this the client calls `RequestStream.CompleteAsync()`, it then waits for `ResponseStream.MoveNext()` to return false.

**Comment by istvanek** [comment]
I will try Your recommendation of calling ResponseStream.MoveNext() after RequestStream.CompleteAsync().
Thank You for Your investigation.

**Comment by istvanek** [comment]
Another case when I see the same exception on server is when I do long time duplex streaming.
Eg. client (both Google and Microsoft one) streams what it is interested in to be notified and server (Microsoft one) notifies client.
In this case client runs ResponseStream.MoveNext() to consume notifications and after some time (10s min) it still waits on that ResponseStream.MoveNext() call without any errors, but the server call end with the error
System.IO.IOException: The client reset the request stream
and effectively stops notifying the client which does not know the call has ended because it still waits on ResponseStream.MoveNext() result.
To workaround this problem, I have used inactivity timeout as CancellationToken passed by client to ResponseStream.MoveNext() and when it occures, I recall to server to continue.
This approach can be used in some cases, but not in all, because some notifications may be lost and there must be another way to get missed notifications to fix it.

**Comment by istvanek** [comment]
We use keep alive as I have written, but it does not heal the problem which keeps client waiting on `ResponseStream.MoveNext()` even when server has already ended the call with `The client reset the request stream` error.
The only workaround for us is using timeout cancellations to recall now.

**Comment by JamesNK** [comment]
If the server has received a notification that the client has killed the RPC call, the RPC call shouldn't hang on the client.

Can you provide a repro of this?

**Comment by istvanek** [comment]
I will try to put a repro together, but it will take more time, because I am working on something else...

**Comment by JamesNK** [comment]
Hi. We're closing this issue as we have heard no response from you for some time now. If you have more details and are encountering this issue please add a new reply and re-open the issue.

**Comment by istvanek** [comment]
@JamesNK sure.
I will create another issue for the last case when I will get more time for replication sample.
Thank You.
